BUG: error Python 3.13 free-threaded `RuntimeError: Identity cache already includes the item.`

### Describe the issue:

When using simple numpy operations in multiple threads, you sometimes get `RuntimeError: Identity cache already includes the item.`

Looks like this had already been noticed in some cases by @ogrisel with Python 3.9 nogil in https://github.com/numpy/numpy/issues/21337 2 years ago, although according to https://github.com/colesbury/nogil/issues/48#issuecomment-1098438898 it was fixed in the nogil Numpy https://github.com/colesbury/numpy/commits/v1.22.3-nogil. Maybe some changes have not been fully upstreamed?

cc @rgommers and @ngoldbaum.

I run the script below like this, which triggers the error eventually:
```bash
for i in $(seq 1 100); do PYTHON_GIL=0 python test-numpy.py; done     
```

Note that there is nothing special about `a.sum` below, you can trigger the same issue with `np.max`, `a != 1`, etc ...

### Reproduce the code example:

```python
import numpy as np
import threading, queue

a = np.ones(1000)

def f(*args, **kwargs):
    return a.sum()

q = queue.Queue()

def worker():
    while True:
        item = q.get()
        f(item)
        q.task_done()

for i in range(2):
    threading.Thread(target=worker, daemon=True, name=f"worker-{i}").start()

for item in range(100000):
    q.put(item)

q.join()
print("done")
```


### Error message:

```shell
Exception in thread worker-1:
Traceback (most recent call last):
  File "/opt/python-free-threading-release/lib/python3.13/threading.py", line 1039, in _bootstrap_inner
    self.run()
    ~~~~~~~~^^
  File "/opt/python-free-threading-release/lib/python3.13/threading.py", line 990, in run
    self._target(*self._args, **self._kwargs)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/test-numpy2.py", line 15, in worker
    f(item)
    ~^^^^^^
  File "/tmp/test-numpy2.py", line 7, in f
    return a.sum()
           ~~~~~^^
  File "/home/lesteve/venvs/python3.13-free-threading-no-debug/lib/python3.13/site-packages/numpy/_core/_methods.py", line 52, in _sum
    return umr_sum(a, axis, dtype, out, keepdims, initial, where)
RuntimeError: Identity cache already includes the item.
```


### Python and NumPy Versions:

```
2.1.0.dev0+git20240611.ee73d0d
3.13.0b1+ experimental free-threading build (heads/3.13:6892b400dc, May 24 2024, 17:10:10) [GCC 14.1.1 20240507]
```

### Runtime Environment:

```
[{'numpy_version': '2.1.0.dev0+git20240524.1f73030',
  'python': '3.13.0b1+ experimental free-threading build '
            '(heads/3.13:6892b400dc, May 24 2024, 17:10:10) [GCC 14.1.1 '
            '20240507]',
  'uname': uname_result(system='Linux', node='thinkpad', release='6.9.2-arch1-1', version='#1 SMP PREEMPT_DYNAMIC Sun, 26 May 2024 01:30:29 +0000', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL',
                                    'AVX512_SPR']}},
 {'architecture': 'Haswell',
  'filepath': '/usr/lib/libopenblas.so.0.3',
  'internal_api': 'openblas',
  'num_threads': 12,
  'prefix': 'libopenblas',
  'threading_layer': 'openmp',
  'user_api': 'blas',
  'version': '0.3.27'},
 {'filepath': '/usr/lib/libgomp.so.1.0.0',
  'internal_api': 'openmp',
  'num_threads': 12,
  'prefix': 'libgomp',
  'user_api': 'openmp',
  'version': None}]
```

### Context for the issue:

This was actually observed first in scikit-learn test suite locally and in one [CI build](https://github.com/lesteve/scikit-learn/actions/runs/9501475134/job/26187089099)

<details>

```
  =================================== FAILURES ===================================
  ___________________ test_sgd_multiclass_njobs[SGDClassifier] ___________________
  joblib.externals.loky.process_executor._RemoteTraceback: 
  """
  Traceback (most recent call last):
    File "/tmp/tmp.QFChQugvDY/venv/lib/python3.13/site-packages/joblib/_utils.py", line 72, in __call__
      return self.func(**kwargs)
             ~~~~~~~~~^^^^^^^^^^
    File "/tmp/tmp.QFChQugvDY/venv/lib/python3.13/site-packages/joblib/parallel.py", line 598, in __call__
      return [func(*args, **kwargs)
              ~~~~^^^^^^^^^^^^^^^^^
    File "/tmp/tmp.QFChQugvDY/venv/lib/python3.13/site-packages/sklearn/utils/parallel.py", line 127, in __call__
      return self.function(*args, **kwargs)
             ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
    File "/tmp/tmp.QFChQugvDY/venv/lib/python3.13/site-packages/sklearn/linear_model/_stochastic_gradient.py", line 436, in fit_binary
      y_i, coef, intercept, average_coef, average_intercept = _prepare_fit_binary(
                                                              ~~~~~~~~~~~~~~~~~~~^
          est, y, i, input_dtye=X.dtype
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
    File "/tmp/tmp.QFChQugvDY/venv/lib/python3.13/site-packages/sklearn/linear_model/_stochastic_gradient.py", line 343, in _prepare_fit_binary
      y_i[y != est.classes_[i]] = -1.0
          ^^^^^^^^^^^^^^^^^^^^
  RuntimeError: Identity cache already includes the item.
  """
  
  The above exception was the direct cause of the following exception:
  
  klass = <function SGDClassifier at 0x5de0cddbb80>
  
      @pytest.mark.parametrize("klass", [SGDClassifier, SparseSGDClassifier])
      def test_sgd_multiclass_njobs(klass):
          # Multi-class test case with multi-core support
  >       clf = klass(alpha=0.01, max_iter=20, n_jobs=2).fit(X2, Y2)
  
  ../venv/lib/python3.13/site-packages/sklearn/linear_model/tests/test_sgd.py:685: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  ../venv/lib/python3.13/site-packages/sklearn/base.py:1514: in wrapper
      return fit_method(estimator, *args, **kwargs)
  ../venv/lib/python3.13/site-packages/sklearn/linear_model/_stochastic_gradient.py:938: in fit
      return self._fit(
  ../venv/lib/python3.13/site-packages/sklearn/linear_model/_stochastic_gradient.py:725: in _fit
      self._partial_fit(
  ../venv/lib/python3.13/site-packages/sklearn/linear_model/_stochastic_gradient.py:649: in _partial_fit
      self._fit_multiclass(
  ../venv/lib/python3.13/site-packages/sklearn/linear_model/_stochastic_gradient.py:804: in _fit_multiclass
      result = Parallel(
  ../venv/lib/python3.13/site-packages/sklearn/utils/parallel.py:65: in __call__
      return super().__call__(iterable_with_config)
  ../venv/lib/python3.13/site-packages/joblib/parallel.py:2007: in __call__
      return output if self.return_generator else list(output)
  ../venv/lib/python3.13/site-packages/joblib/parallel.py:1650: in _get_outputs
      yield from self._retrieve()
  ../venv/lib/python3.13/site-packages/joblib/parallel.py:1754: in _retrieve
      self._raise_error_fast()
  ../venv/lib/python3.13/site-packages/joblib/parallel.py:1789: in _raise_error_fast
      error_job.get_result(self.timeout)
  ../venv/lib/python3.13/site-packages/joblib/parallel.py:745: in get_result
      return self._return_or_raise()
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  self = <joblib.parallel.BatchCompletionCallBack object at 0x5de13e58d10>
  
      def _return_or_raise(self):
          try:
              if self.status == TASK_ERROR:
  >               raise self._result
  E               RuntimeError: Identity cache already includes the item.
  
  ../venv/lib/python3.13/site-packages/joblib/parallel.py:763: RuntimeError
```

</details>

--- a/numpy/_core/src/common/npy_hashtable.c
+++ b/numpy/_core/src/common/npy_hashtable.c
@@ -234,7 +234,7 @@ PyArrayIdentityHash_SetItem(PyArrayIdentityHash *tb,
         if (tb_item[0] != NULL && tb_item[0] != value && !replace) {
             UNLOCK_TABLE(tb);
             PyErr_SetString(PyExc_RuntimeError,
-                    "Identity cache already includes the item.");
+                    "Identity cache already includes an item with this key.");
             return -1;
         }
         tb_item[0] = value;

--- a/numpy/_core/src/common/npy_hashtable.c
+++ b/numpy/_core/src/common/npy_hashtable.c
@@ -210,10 +210,13 @@ _resize_if_necessary(PyArrayIdentityHash *tb)
  * @param value Normally a Python object, no reference counting is done.
  *        use NULL to clear an item.  If the item does not exist, no
  *        action is performed for NULL.
- * @param replace If 1, allow replacements.
+ * @param replace If 1, allow replacements. If replace is 0 an error is raised
+ *        if the stored value is different from the value to be cached. If the
+ *        value to be cached is identical to the stored value, the value to be
+ *        cached is ignored and no error is raised.
  * @returns 0 on success, -1 with a MemoryError or RuntimeError (if an item
- *        is added which is already in the cache).  The caller should avoid
- *        the RuntimeError.
+ *        is added which is already in the cache and replace is 0).  The
+ *        caller should avoid the RuntimeError.
  */
 NPY_NO_EXPORT int
 PyArrayIdentityHash_SetItem(PyArrayIdentityHash *tb,
@@ -228,7 +231,7 @@ PyArrayIdentityHash_SetItem(PyArrayIdentityHash *tb,
 
     PyObject **tb_item = find_item(tb, key);
     if (value != NULL) {
-        if (tb_item[0] != NULL && !replace) {
+        if (tb_item[0] != NULL && tb_item[0] != value && !replace) {
             UNLOCK_TABLE(tb);
             PyErr_SetString(PyExc_RuntimeError,
                     "Identity cache already includes the item.");

--- a/numpy/_core/tests/test_hashtable.py
+++ b/numpy/_core/tests/test_hashtable.py
@@ -23,8 +23,3 @@ def test_identity_hashtable(key_length, length):
 
     res = identityhash_tester(key_length, keys_vals, replace=True)
     assert res is expected
-
-    # check that ensuring one duplicate definitely raises:
-    keys_vals.insert(0, keys_vals[-2])
-    with pytest.raises(RuntimeError):
-        identityhash_tester(key_length, keys_vals)

--- a/numpy/_core/tests/test_hashtable.py
+++ b/numpy/_core/tests/test_hashtable.py
@@ -23,3 +23,13 @@ def test_identity_hashtable(key_length, length):
 
     res = identityhash_tester(key_length, keys_vals, replace=True)
     assert res is expected
+
+    if length == 1:
+        return
+
+    # add a new item with a key that is already used and a new value, this
+    # should error if replace is False, see gh-26690
+    new_key = (keys_vals[1][0], object())
+    keys_vals[0] = new_key
+    with pytest.raises(RuntimeError):
+        identityhash_tester(key_length, keys_vals)

--- a/numpy/_core/tests/test_multithreading.py
+++ b/numpy/_core/tests/test_multithreading.py
@@ -1,4 +1,5 @@
 import concurrent.futures
+import threading
 
 import numpy as np
 import pytest
@@ -30,12 +31,28 @@ def test_parallel_randomstate_creation():
 def test_parallel_ufunc_execution():
     # if the loop data cache or dispatch cache are not thread-safe
     # computing ufuncs simultaneously in multiple threads leads
-    # to a data race
+    # to a data race that causes crashes or spurious exceptions
     def func():
         arr = np.random.random((25,))
         np.isnan(arr)
 
     run_threaded(func, 500)
+
+    # see gh-26690
+    NUM_THREADS = 50
+
+    b = threading.Barrier(NUM_THREADS)
+
+    a = np.ones(1000)
+
+    def f():
+        b.wait()
+        return a.sum()
+
+    threads = [threading.Thread(target=f) for _ in range(NUM_THREADS)]
+
+    [t.start() for t in threads]
+    [t.join() for t in threads]
 
 def test_temp_elision_thread_safety():
     amid = np.ones(50000)

